src/dbManager.js

```diff
@@ -0,0 +1,136 @@
+const BN = require('bn.js')
+const level = require('level-mem')
+const rlp = require('rlp')
+const Block = require('ethereumjs-block')
+const Cache = require('./cache')
+const {
+  headsKey,
+  headHeaderKey,
+  headBlockKey,
+  hashToNumberKey,
+  numberToHashKey,
+  tdKey,
+  bodyKey,
+  headerKey
+} = require('./util')
+
+module.exports = class DBManager {
+  constructor (db, common) {
+    this._db = db
+    this._common = common
+    this._cache = {
+      td: new Cache({ max: 1024 }),
+      header: new Cache({ max: 512 }),
+      body: new Cache({ max: 256 }),
+      numberToHash: new Cache({ max: 2048 }),
+      hashToNumber: new Cache({ max: 2048 })
+    }
+  }
+
+  getHeads () {
+    return this.get(headsKey, { valueEncoding: 'json' })
+  }
+
+  getHeadHeader () {
+    return this.get(headHeaderKey)
+  }
+
+  getHeadBlock () {
+    return this.get(headBlockKey)
+  }
+
+  async get (key, opts = {}) {
+    const dbOpts = {
+      keyEncoding: opts.keyEncoding || 'binary',
+      valueEncoding: opts.valueEncoding || 'binary'
+    }
+
+    if (opts.cache) {
+      if (!this._cache[opts.cache]) {
+        throw new Error(`Invalid cache: ${opts.cache}`)
+      }
+
+      let value = this._cache[opts.cache].get(key)
+      if (!value) {
+        value = await this._db.get(key, dbOpts)
+        this._cache[opts.cache].set(key, value)
+      }
+
+      return value
+    }
+
+    return this._db.get(key, dbOpts)
```

Comment:
This is also still operating on `this._db`?

Reply:
Well, its the `get` method, so... I didn't know how to avoid using `this._db.get` :sweat_smile: Open to ideas.

Reply:
Ah, sorry, put this somewhere else due to all this comments interrupting the code flow. ARgh. 🤣